be/src/vec/functions/array/function_array_aggregation.cpp

```diff
@@ -0,0 +1,303 @@
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+// This file is copied from
+// https://github.com/ClickHouse/ClickHouse/blob/master/src/Functions/array/arrayAggregation.cpp
+// and modified by Doris
+
+#include <condition_variable>
+#include <type_traits>
+
+#include "vec/common/arithmetic_overflow.h"
+#include "vec/core/types.h"
+#include "vec/data_types/data_type_array.h"
+#include "vec/data_types/data_type_decimal.h"
+#include "vec/data_types/data_type_number.h"
+#include "vec/functions/array/function_array_mapped.h"
+#include "vec/functions/simple_function_factory.h"
+
+namespace doris {
+namespace vectorized {
+
+#define RETURN_TRUE_WITH_RUNTIME_ERROR(status, msg) \
+    do {                                            \
+        status = Status::RuntimeError(msg);         \
+        return true;                                \
+    } while (0);
+
+enum class AggregateOperation { min, max, sum, average, product };
+
+/**
+ * During array aggregation we derive result type from operation.
+ * For array min or array max we use array element as result type.
+ * For array average we use Float64.
+ * For array sum for for big integers, we use same type representation, decimal numbers we use Decimal128,
+ * for floating point numbers Float64, for numeric unsigned Int64, and for numeric signed UInt64.
+ */
+
+template <typename ArrayElement, AggregateOperation operation>
+struct ArrayAggregateResultImpl;
+
+template <typename ArrayElement>
+struct ArrayAggregateResultImpl<ArrayElement, AggregateOperation::min> {
+    using Result = ArrayElement;
+};
+
+template <typename ArrayElement>
+struct ArrayAggregateResultImpl<ArrayElement, AggregateOperation::max> {
+    using Result = ArrayElement;
+};
+
+template <typename ArrayElement>
+struct ArrayAggregateResultImpl<ArrayElement, AggregateOperation::average> {
+    using Result = Float64;
+};
+
+template <typename ArrayElement>
+struct ArrayAggregateResultImpl<ArrayElement, AggregateOperation::product> {
+    using Result = Float64;
+};
+
+template <typename ArrayElement>
+struct ArrayAggregateResultImpl<ArrayElement, AggregateOperation::sum> {
+    using Result = std::conditional_t<
+            IsDecimalNumber<ArrayElement>, Decimal128,
+            std::conditional_t<
+                    IsFloatNumber<ArrayElement>, Float64,
+                    std::conditional_t<std::is_same_v<ArrayElement, Int128>, Int128, Int64>>>;
+};
+
+template <typename ArrayElement, AggregateOperation operation>
+using ArrayAggregateResult = typename ArrayAggregateResultImpl<ArrayElement, operation>::Result;
+
+template <AggregateOperation aggregate_operation>
+struct ArrayAggregateImpl {
+    using column_type = ColumnArray;
+    using data_type = DataTypeArray;
+
+    static DataTypePtr getReturnType(const DataTypePtr& expression_return,
```

Comment:
for `average` why not use `ArrayAggregateResultImpl`

Reply:
Refined.